rust/candid/src/types/mod.rs

```diff
@@ -17,31 +17,66 @@ pub mod principal;
 pub mod reference;
 pub mod reserved;
 
-pub trait CandidType {
-    // memoized type derivation
-    fn ty() -> Type {
+pub trait IdlSerialize {
+    // only serialize the value encoding
+    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
+    where
+        S: Serializer;
+}
+
+pub trait CandidTyping<Cache = StaticCache>: IdlSerialize {
+    fn ty() -> Type
+    where
+        Cache: Default,
+    {
+        Self::ty_from_cache(&mut Default::default())
+    }
+
+    fn ty_from_cache(c: &mut Cache) -> Type;
+}
+
+impl<T: CandidType + ?Sized, Cache: CandidTypeCache> CandidTyping<Cache> for T {
+    fn ty_from_cache(c: &mut Cache) -> Type {
         let id = Self::id();
-        if let Some(t) = self::internal::find_type(&id) {
-            match t {
-                Type::Unknown => Type::Knot(id),
-                _ => t,
+        match c.find_type(&id) {
+            Some(Type::Unknown) => Type::Knot(id),
+            Some(t) => t,
+            None => {
+                c.add_type(id.clone(), Type::Unknown);
+                let t = Self::_ty(c);
+                c.add_type(id, t.clone());
+                t
             }
-        } else {
-            self::internal::env_add(id.clone(), Type::Unknown);
-            let t = Self::_ty();
-            self::internal::env_add(id.clone(), t.clone());
-            self::internal::env_id(id, t.clone());
-            t
         }
     }
+}
+
+pub trait CandidTypeCache {
+    fn find_type(&self, id: &TypeId) -> Option<Type>;
+    fn add_type(&mut self, id: TypeId, ty: Type);
+}
+
+#[derive(Default)]
+pub struct StaticCache;
+
+impl CandidTypeCache for StaticCache {
+    fn find_type(&self, id: &TypeId) -> Option<Type> {
+        self::internal::find_type(&id)
+    }
+    fn add_type(&mut self, id: TypeId, ty: Type) {
+        if ty != Type::Unknown {
+            self::internal::env_id(id.clone(), ty.clone());
+        }
+        self::internal::env_add(id, ty);
+    }
+}
+
+pub trait CandidType: IdlSerialize {
```

Comment:
What's the difference between `CandidType` and `CandidTyping`?

Reply:
Basically I split `CandidTyping` off from `CandidType`, leaving all the implementation side `_ty` stuff in `CandidType`, but taking the caller side `ty` into `CandidTyping`. This allows `IDLValue` to override the `CandidTyping` without needing the `unreachable!` implementations.

The relationship between them is kind of like `From` and `Into`, except with the extra layer of memoization `CandidTyping` is responsible for invoking.

Reply:
`IDLValue` is not supposed to be used for serializing native Rust values. The `unreachable` is intentional, so that users know it's an anti-pattern. Splitting the trait just to avoid this `unreachable` seems overkill. Also since `0.7.0`, `IDLValue::ty()` returns `Type::Unknown` instead of `unreachable`.